cloud/services/computes/droplets.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package computes
+
+import (
+	"encoding/base64"
+	"net/http"
+	"strconv"
+
+	"github.com/digitalocean/godo"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	infrav1 "sigs.k8s.io/cluster-api-provider-digitalocean/api/v1alpha2"
+	"sigs.k8s.io/cluster-api-provider-digitalocean/cloud/scope"
+)
+
+// GetDroplet get a droplet instance
+func (s *Service) GetDroplet(id string) (*godo.Droplet, error) {
+	if id == "" {
+		s.scope.Info("DOMachine does not have an instance id")
+		return nil, nil
+	}
+
+	s.scope.V(2).Info("Looking for instance by id", "instance-id", id)
+	dropletID, err := strconv.Atoi(id)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse instance id with id %q", id)
+	}
+
+	droplet, res, err := s.scope.Droplets.Get(s.ctx, dropletID)
+	if err != nil {
+		if res.StatusCode == http.StatusNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	if droplet != nil {
+		return droplet, nil
+	}
+	return nil, nil
+}
+
+// CreateDroplet create a droplet instance.
+func (s *Service) CreateDroplet(scope *scope.MachineScope) (*godo.Droplet, error) {
+	s.scope.V(2).Info("Creating an instance for a machine")
+
+	bootstrapData, err := base64.StdEncoding.DecodeString(*scope.Machine.Spec.Bootstrap.Data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode bootstrap data")
+	}
+
+	clusterName := infrav1.DOSafeName(s.scope.Name())
+	instanceName := infrav1.DOSafeName(scope.Name())
+	volumeID := scope.InstanceVolumeRef().ResourceID
+
+	image, err := s.GetImage(scope)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed getting image")
+	}
+
+	sshkeys := []godo.DropletCreateSSHKey{}
+	for _, v := range scope.DOMachine.Spec.SSHKeys {
+		keys, err := s.GetSSHKey(v)
+		if err != nil {
+			return nil, err
+		}
+		sshkeys = append(sshkeys, godo.DropletCreateSSHKey{
+			ID:          keys.ID,
+			Fingerprint: keys.Fingerprint,
+		})
+	}
```

Comment:
This approach has a downside if no SSH key is provided. In that case, DigitalOcean would e-mail the root password for the instance. This can be a problem from the security aspect, as well as, it can get quite spammy over time.

In the `v1a1`, if no SSH keys were provided, we were generating a random key and we would use that one to avoid emailing the root password. Should we take the same approach here? How other providers are handling this?

Reply:
yep this approach different from the `v1alpha1` type.

but in this approach, i make the `SSHKeys` field as required. so users must provided one or more ssh key.

also the different from the `v1alpha1` type, the `SSHKeys` field is not a public key, but `ssh fingerprint` or `ssh id` that has been registered on DigitalOcean account, and will be error (machine will never created) when these values is not valid or not found in the DigitalOcean account

```go
// SSHKeys is the ssh key id or fingerprint to attach in DigitalOcean droplet.
// It must be available on DigitalOcean account. See https://developers.digitalocean.com/documentation/v2/#list-all-keys
SSHKeys []intstr.IntOrString `json:"sshKeys"`
```

if the value is `int` it will be considered as `ssh id` and if it is a `string` it will be considered as `ssh fingerprint`

i thought it's more secure rather than generating new `ssh-key` where users (often) don't realize it. IMHO

one thing i missed, no failure when these values are not valid or not found in the DigitalOcean account. will be fixed soon.

as far as i know, some other providers (CAPA and CAPG) allows us to make a machine without an ssh key. but CAPA has a similar approach with this, they use `sshKeyName` field but no required.

Reply:
when applying/creating `DOMachine` if the `SSHKeys` field is not provided will be given error like this

```
The DOMachine "capdo-cluster-controlplane-0" is invalid: spec.sshKeys: Invalid value: "null": spec.sshKeys in body must be of type array: "null"
```

and when `SSHKeys` just contain an empty array, it will be given error

```
The DOMachine "capdo-cluster-controlplane-0" is invalid: spec.sshKeys: Invalid value: "null": spec.sshKeys in body must be of type integer,string: "null"
```

and when provided `SSHKeys` is not found in the DigitalOcean account, the machine creating will be failed. and give some error information. something like 

```
$ kubectl describe domachine.infrastructure.cluster.x-k8s.io/capdo-cluster-controlplane-0
...
...
Status:
  Error Message:  Failed to create droplet instance for DOMachine default/capdo-cluster-controlplane-0: GET https://api.digitalocean.com/v2/account/keys/96:56:4d:a9:df:24:85:62:b1:77:c5:4b:ea:6d:9d:aa: 404 The resource you were accessing could not be found.
  Error Reason:   CreateError
Events:
  Type     Reason                 Age   From                  Message
  ----     ------                 ----  ----                  -------
  Warning  InstanceCreatingError  18s   domachine-controller  Failed to create droplet instance for DOMachine default/capdo-cluster-controlplane-0: GET https://api.digitalocean.com/v2/account/keys/96:56:4d:a9:df:24:85:62:b1:77:c5:4b:ea:6d:9d:aa: 404 The resource you were accessing could not be foun
```